lib/directory.coffee

```diff
@@ -163,34 +165,63 @@ class Directory
       names = []
     names.sort(new Intl.Collator(undefined, {numeric: true, sensitivity: "base"}).compare)
 
-    files = []
-    directories = []
-
+    namePromises = []
     for name in names
-      fullPath = path.join(@path, name)
-      continue if @isPathIgnored(fullPath)
-
-      stat = fs.lstatSyncNoException(fullPath)
-      symlink = stat.isSymbolicLink?()
-      stat = fs.statSyncNoException(fullPath) if symlink
-
-      if stat.isDirectory?()
-        if @entries.hasOwnProperty(name)
-          # push a placeholder since this entry already exists but this helps
-          # track the insertion index for the created views
-          directories.push(name)
-        else
-          expansionState = @expansionState.entries[name]
-          directories.push(new Directory({name, fullPath, symlink, expansionState, @ignoredPatterns}))
-      else if stat.isFile?()
-        if @entries.hasOwnProperty(name)
-          # push a placeholder since this entry already exists but this helps
-          # track the insertion index for the created views
-          files.push(name)
-        else
-          files.push(new File({name, fullPath, symlink, realpathCache}))
-
-    @sortEntries(directories.concat(files))
+      # throw this in a closure so we don't have scoping issues
+      # async/await would make this more pleasant when this is
+      # inevitably ported to ES6
+      f = =>
+        localName = name
+        fullPath = path.join(@path, localName)
+
+        return @isPathIgnored(fullPath).then (isIgnored) =>
+          return if isIgnored
+
+          stat = fs.lstatSyncNoException(fullPath)
+          symlink = stat.isSymbolicLink?()
+          stat = fs.statSyncNoException(fullPath) if symlink
+          if stat.isDirectory?()
+            if @entries.hasOwnProperty(localName)
+              # push a placeholder since this entry already exists but this helps
+              # track the insertion index for the created views
+              return [localName, 'directory']
```

Comment:
I'm not sure. I wondered the same thing (it's the same way in the sync code) but I didn't fully grok why the placeholder was used at the time and just kept the existing behavior to avoid more breakage. Now that things are green again it's a better time to mess with it.